source/FreeRTOS_IP_Utils.c

```diff
@@ -398,9 +398,9 @@ void vPreCheckConfigs( void )
             {
                 /* This is a 64-bit platform, make sure there is enough space in
                  * pucEthernetBuffer to store a pointer. */
-                configASSERT( ipconfigBUFFER_PADDING >= 14 );
+                configASSERT( ipBUFFER_PADDING >= 14 );
```

Comment:
To be complete, we might as well check `ipBUFFER_PADDING` in case of a 32-bit platform:
~~~
    #if( UINTPTR_MAX > 0xFFFFFFFF )
    {
        /* This is a 64-bit platform, make sure there is enough space in
         * pucEthernetBuffer to store a pointer. */
        configASSERT( ipBUFFER_PADDING >= 14 );
    }
    #else
    {
        /* This is a 32-bit platform. */
        configASSERT( ipBUFFER_PADDING >= 10 );
    }
    #endif
    /* And it must have this strange alignment: */
    configASSERT( ( ( ( ipBUFFER_PADDING ) + 2 ) % 4 ) == 0 );
~~~

and use the static check for word size: `#if( UINTPTR_MAX > 0xFFFFFFFF )`

Reply:
@htibosch I'm still a bit confused. How are the minimums of 10 and 14 bytes calculated?

Reply:
> I'm still a bit confused. How are the minimums of 10 and 14 bytes calculated?

That is a difficult story, have a look [here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/dd42247cc0fd3fdef51b0fa902bb22923a525d83/source/include/FreeRTOS_IP_Private.h#L452).

`ipIP_TYPE_OFFSET` is the offset of 6 bytes before the actual packet where the IP version is written.
~~~
address -10 pointer to the owning NetworkBuffer
address  -9 pointer
address  -8 pointer
address  -7 pointer
address  -6 IP-type, either 0x40 or 0x60
address  -5 n.a.
address  -4 n.a.
address  -3 n.a.
address  -2 n.a.
address  -1 n.a.
address   0 Ethernet header
~~~

The IP-type is read by `pxUDPPayloadBuffer_to_NetworkBuffer()` :

The problem is that the library passes pointers to the UDP payload (used for zero-copy `recvfrom()`/`sendto()`).
At some point, we have a payload and we want to know the start of the packet. Therefore we need to know if it is an IPv4 or an IPv6 packet. The payload doesn't tell anything about this, it is just a user message.

For an IPv6 packet, the type is stored in the IP-header in the field `ucVersionTrafficClass`, which has the value 0x60. This field is located 48 bytes before the payload.
For an IPv4 packet, at a position of 48 bytes before the payload, the value `0x40` will be stored. That is a position **before** the Ethernet header, marked here above as "IP-type".

So that means that `ipBUFFER_PADDING` must be at least 10 bytes long. And as pointers on a 64-bit platform are 8 bytes long, the minimum length of `ipBUFFER_PADDING` is 14 or longer.

Reply:
Maybe a better way to visualise the message buffer header is as follows:
~~~c
#define ipTYPE_IPv4         ( 0x40U )
#define ipTYPE_IPv6         ( 0x60U )

struct NetworkBuffer {
    NetworkBufferDescriptor_t * pxOwner;
    uint8_t ucIPType;  /* Either 'ipTYPE_IPv4' or 'ipTYPE_IPv6' */
    uint8_t ucFiller[ 5 ];
    uint8_t ucEthernetHeader[ 14 ];
    uint8_t ucIPHeader[ 20 ];  /* Either 20 or 40 */
};
~~~
The field `ucIPHeader` must be 4-byte aligned.